### PR TITLE
Geoff/group lasso

### PR DESCRIPTION
**Pull request by GeoffNN**
First attempt at implementing a penalized GroupLasso. The `__call__` method seems correct. The `prox` still needs some work.

### CODE REVIEW

**Comment by fabianp** [comment]
perhaps a good test would be to implement the prox for the L1 and then check that Group Lasso is the same as L1 when groups are of size one

**Comment by GeoffNN** [comment]
@fabianp Done, tests pass. Comments welcome :)

**Comment by fabianp** [comment]
amazing, will take a look ASAP

File: chop/penalties.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+                                       for g in self.groups])
+        else:
+            group_norms = torch.stack([torch.linalg.norm(x[(Ellipsis,) + tuple(zip(*g))],
+                                                         dim=-1)
+                                       for g in self.groups])
+
+        return self.alpha * group_norms.sum(dim=0)
+
+    @torch.no_grad()
+    def prox(self, x, step_size=None):
```
Comment (fabianp) [inline chop/penalties.py:59]: Add docstring with at least reference to where this prox is derived

File: chop/penalties.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        out = x.detach().clone()
+        if type(step_size) == float:
+            step_size *= torch.ones(x.size(0))
+
+        if x.dim() == 2:
```
Comment (fabianp) [inline chop/penalties.py:64]: would be nice to have a comment on why this distinction of cases

Comment (fabianp) [reply chop/penalties.py:64]: and why it's not present in the L1 prox

File: chop/penalties.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+    def prox(self, x, step_size=None):
+        """Prox operator for L1 norm. This is given by soft-thresholding."""
+        return torch.sign(x) * F.relu(abs(x) - self.alpha * step_size)
+
+
+class GroupL1:
+    """
+    Group LASSO penalty. Batch-wise function.
+    """
```
Comment (fabianp) [inline chop/penalties.py:34]: add a description of the parameters in the init (specially groups since its non-trivial)

Comment (fabianp) [reply chop/penalties.py:34]: even better: add an example. I'm reading the tests and its non-trivial to understand how the groups are represented

Comment (GeoffNN) [reply chop/penalties.py:34]: Working on adding an adversarial-patch image example. I'll keep it in this branch.

The groups are an iterable of iterables. They're transformed in an iterable of torch.Tensors used for indexing. I can't directly use a torch.Tensor of higher dim because the groups might be of different sizes.

File: chop/penalties.py (hunk 2 above)
Comment (GeoffNN) [reply chop/penalties.py:64]: Right, I think I need to add this to the L1 prox for consistency with the `constraints` module

File: chop/penalties.py (hunk 3 above)
Comment (GeoffNN) [reply chop/penalties.py:61]: Added explanations in the docstring.
